src/System.CodeDom/tests/CodeGenerationTests.cs

```diff
@@ -123,44 +123,44 @@ protected void AssertEqual(CodeObject c, string expected)
 
             // Serialize and deserialize the CodeObject, and make sure code generated for it
             // is the same as the original.
-            CodeObject clone = BinaryFormatterHelpers.Clone(c);
+            CType clone = BinaryFormatterHelpers.Clone(c);
             string cloneCode = GenerateCode(clone, provider);
             Assert.Equal(code, cloneCode);
         }
 
-        private static string GenerateCode(CodeObject c, CodeDomProvider provider)
+        private static string GenerateCode<CType>(CType c, CodeDomProvider provider) where CType : CodeObject
         {
             var sb = new StringBuilder();
             var writer = new StringWriter(sb);
             var options = new CodeGeneratorOptions();
 
             if (c is CodeStatement)
             {
-                provider.GenerateCodeFromStatement((CodeStatement)c, writer, options);
+                provider.GenerateCodeFromStatement(c as CodeStatement, writer, options);
```

Comment:
... I think I was just trying to be too clever with things here, in general, and reverted the typing/casting changes.  I left the switch to the exception alone.

(well, I had to use this form, because the use of generics meant the compiler couldn't see a path to the type, since `c` was a derived type of one of those mentioned, but obviously not all of them.  Took me too long to figure that out, though.)